lib/galaxy/tools/error_reports/plugins/gitlab.py

```diff
@@ -0,0 +1,144 @@
+"""The module describes the ``gitlab`` error plugin plugin."""
+from __future__ import absolute_import
+
+import logging
+
+from galaxy.tools.errors import EmailErrorReporter
+from galaxy.util import string_as_bool, unicodify
+from . import ErrorPlugin
+
+log = logging.getLogger(__name__)
+
+
+class GitLabPlugin(ErrorPlugin):
+    """Send error report to GitLab.
+    """
+    plugin_type = "gitlab"
+
+    def __init__(self, **kwargs):
+        self.app = kwargs['app']
+        self.redact_user_details_in_bugreport = self.app.config.redact_user_details_in_bugreport
+        self.verbose = string_as_bool(kwargs.get('verbose', False))
+        self.user_submission = string_as_bool(kwargs.get('user_submission', False))
+
+        # GitLab settings
+        self.gitlab_base_url = kwargs.get('gitlab_base_url', 'https://gitlab.com')
+        self.gitlab_default_repo_owner = kwargs.get('gitlab_default_repo_owner', False)
+        self.gitlab_default_repo_name = kwargs.get('gitlab_default_repo_name', False)
+
+        try:
+            import gitlab
+            import requests
+            import os
+
+            session = requests.Session()
+            session.proxies = {
+                'https': os.environ.get('https_proxy'),
+                'http': os.environ.get('http_proxy'),
+            }
+            self.gitlab = gitlab.Gitlab(
+                # Allow running against GL enterprise deployments
+                kwargs.get('gitlab_base_url', 'https://gitlab.com'),
+                private_token=kwargs['gitlab_private_token'],
+                session=session
+            )
+            self.gitlab.auth()
+
+        except ImportError:
+            log.error("Please install python-gitlab to submit bug reports to gitlab")
+            self.gitlab = None
+
+    def submit_report(self, dataset, job, tool, **kwargs):
+        """Submit the error report to GitLab
+        """
+        log.info(self.gitlab)
+
+        if self.gitlab:
+            # Import the necessary libraries
+            import sys
+            import requests
+            if sys.version_info[0] < 3:
+                import urllib as urllib
+                import urlparse as urlparse
+            else:
+                import urllib.parse as urllib
+                urlparse = urllib
+
+            log.info(job.tool_id)
+            log.info(job.tool_version)
+            log.info(tool.tool_shed)
+
+            # Determine the ToolShed
+            ts_url_request = requests.get('http://' + str(tool.tool_shed))
```

Comment:
```suggestion
            ts_url_request = requests.get('https://' + str(tool.tool_shed))
```

Reply:
@erasche the idea behind this piece of code is to get Python to determine whether the website is HTTP or HTTPS. As not all ToolSheds will have HTTPS, this piece of code actually detects the HTTP redirect. So for instance for the testtoolshed, it goes to HTTPS. It just removes the requirement for needing a SSL certificate on the ToolShed for people their ToolShed deployments.

Reply:
I would expect the `tool.tool_shed` to be correct, but I can recognise that if it is written `http` in the tool, while the server actually responds on HTTPS, you don't want to have to update your tool for the bug reporter to work correctly.

OK, sounds fine. (But people should really write the correct tool shed URL in their tool shed repo metadata.)